File: ShellBoost.Samples/ShellBoost.Samples.Overview/SimpleItem.cs
﻿using System;
using ShellBoost.Core;
using ShellBoost.Core.Client;
using ShellBoost.Core.Utilities;
using ShellBoost.Core.WindowsPropertySystem;

namespace ShellBoost.Samples.Overview
{
    public class SimpleItem : ShellItem
    {
        public SimpleItem(ShellFolder parent, string text)
            : base(parent, new StringKeyShellItemId(text))
        {
            // this is needed for icon
            ItemType = IOUtilities.PathGetExtension(text);
            CanCopy = true;
        }

        public override bool TryGetPropertyValue(PropertyKey key, out object value)
        {
            if (key == Core.WindowsPropertySystem.System.PropList.InfoTip)
            {
                value = null;
                return false;
            }

            // dynamic infotip (aka: tooltip)
            if (key == Core.WindowsPropertySystem.System.InfoTipText)
            {
                value = "This is " + DisplayName + ", info created " + DateTime.Now;
                return true;
            }

            return base.TryGetPropertyValue(key, out value);
        }

        // give back some dynamic content
        public override ShellContent GetContent() => new MemoryShellContent(DisplayName + " - this is dynamic content created from .NET " + (Installer.IsNetCore ? "Core" : "Framework") + " at " + DateTime.Now);
    }
}

File: ShellBoost.Samples/ShellBoost.Samples.CppOverview/SimpleItem.cpp
#include "pch.h"
#include "SimpleItem.h"

using namespace ShellBoost::Core::Utilities;

SimpleItem::SimpleItem(ShellFolder^ parent, String^ text)
	:ShellItem(parent, gcnew StringKeyShellItemId(text))
{
	// this is needed for icon
	ItemType = IOUtilities::PathGetExtension(text);
	CanCopy = true;
}

bool SimpleItem::TryGetPropertyValue(WindowsPropertySystem::PropertyKey key, Object^% value)
{
	if (key == WindowsPropertySystem::System::PropList::InfoTip)
	{
		value = nullptr;
		return false;
	}

	// dynamic infotip (aka: tooltip)
	if (key == WindowsPropertySystem::System::InfoTipText)
	{
		value = L"This is " + DisplayName + L", info created " + DateTime::Now;
		return true;
	}

	return ShellItem::TryGetPropertyValue(key, value);
}

ShellContent^ SimpleItem::GetContent()
{
	auto content = gcnew MemoryShellContent(DisplayName + L" - this is dynamic content created from .NET Framework C++/CLI at " + DateTime::Now);
	return content;
}